### DISCUSSION TITLE
v0.8.0-alpha : Rules and Keys

### DISCUSSION THREAD

**Discussion by U-C-S** [U-C-S/Hurl#69] [Announcements]
This is a tough one from user-experience POV, I did what I thought was best as I did dogfooding and ideas from discussion in the linked isssues. "Hurl Settings" is not available yet with this alpha release. 

## Feature Updates
- Add Basic Browser Rules Support, with support for String, Regex, Glob match (#18)
- New Quick Add Rule window to add rules from Hurl mainwindow
- Few Keyboard Shortcuts for copying URl (C), editing URI (E), Quick adding rule (R), Timed browser (T) (#44)
- Updated Hurl's Toolbar layout to get more focus
- misc:
  - (BREAKING) Remove `DisableAcrylic` and `BackgroundRGB` setting options
  - (BREAKING) rename `UseWhiteBorder` to `NoWhiteBorder`
  
## Development Updates
- Migrate to Singleton Pattern to ease dev process. (which is a half-lie and now I can see people are against it)
- Yet an another project structure update to make space for Hurl Settings project
- Revert back to using wpfui repo as git submodule, instead of Nuget package

Please provide any feedback in the discussion thread for this release. Stable release can be expected by the end of April.

There's no installer provided with this alpha release. Copy the exe file into the hurl installed location (default at `C:\Program Files (x86)\Hurl`

**Full Changelog**: https://github.com/U-C-S/Hurl/compare/v0.7.1...v0.8.0-alpha

<hr /><em>This discussion was created from the release <a href='https://github.com/U-C-S/Hurl/releases/tag/v0.8.0-alpha'>v0.8.0-alpha : Rules and Keys</a>.</em>

**Comment by U-C-S** [comment]
Adding new rule can be done from UI by clicking `+` btn in the Hurl main window

You can add new rules into `AutoRoutingRules` property, which itself is a list of Rules Objects. Each rule object should contain 2 properties 
  - `Rules` : it can be a single rule or multiple rules. (rules are prioritiesed, so first rule that matches will trigger the opening of browser)
  - `BrowserName`: The value should be a Browser name that matches one of the `Browsers` names. which will be opened once a rule from the `Rules` property matches. Use `_Hurl` value to open Hurl

Sample Json file for more advanced editing.... 
```json
{
  "Browsers": [
    {
      "Name": "Firefox Nightly",
      "ExePath": "C:\\Program Files\\Firefox Nightly\\firefox.exe"
    },
    {
      "Name": "Google Chrome Dev",
      "ExePath": "C:\\Program Files\\Google\\Chrome Dev\\Application\\chrome.exe"
    }
  ],
  "AutoRoutingRules": [
    {
      "Rules": ["g$*.google.com*"],
      "BrowserName": "Google Chrome Dev"
    },
    {
      "Rules": ["https://github.com/u-c-s", "r$.*open\\.spotify\\.com.*"],
      "BrowserName": "Firefox Nightly"
    }
  ]
}
```

Rule Check Mode
- `Regex`: rule should start with `r$` to activate Regex mode. Use https://regex101.com with (.NET/C# flavor) to test ur rules.
- `Glob`: rule should start with `g$` for hurl to detect glob mode. continue writing ur actual rule after that. Use https://www.digitalocean.com/community/tools/glob to test ur rules
- `String`: This does simple text matching. triggers if ur URL matches exactly with this plain text rule. no prefixes

**Comment by riotrah** [comment]
Thanks for this.

Any chance there's a JSON Schema for the settings?

**Reply by U-C-S** [reply]
Try this one.... This is a first for me to write a json schema

raw json: https://gist.githubusercontent.com/U-C-S/1f3fe5876a3b2ccf0ca5c7e630427185/raw/814c26ef83872a4e2304d68e4583c3dc57276a7e/hurl_settings_schema.json
Gist: https://gist.github.com/U-C-S/1f3fe5876a3b2ccf0ca5c7e630427185

(I will add property descriptions later once stable v0.8 happens....)

**Reply by riotrah** [reply]
I don't know if you're still uncertain about the settings UI, or which aspect of it you're most uncertain about, but I wonder, now that you have some form of schema describing the type and requirements of each settings field, whether initial versions could use dynamically rendered inputs and such from the schema. 
At that point iteration on the UI and UX of the settings could be done separately (and/or with help) from iterations on the configurability of the app itself or other features. Maybe that's not at all useful or relevant to the most difficult part of it for you but would love to help either way.

**Reply by U-C-S** [reply]
You got it right about my uncertainty about Settings UI. Mostly it's because I'm still not familiar with WPF's design patterns which I already use for Hurl. It's versatile enough for Hurl to be possible. But Hurl Settings doesn't need most of those. but also doesn't want to move to different language or runtime so I could share some logic in between them smoothly. (like Browser installed checker, and other future use case).

Seems I found a comprise on that now with recent release of Windows App SDK, which is well maintained but still infant. But is the future of Desktop applications. I could technically use it for Hurl but Hurl needs faster loading times and lightweight and other stuff. But for Hurl Settings I think it's a good option. Installer size might go up a 80mb but it should be easier to develop with for a CRUD like app i.e. Hurl Settings.

TL;DR : Thinking to use different but heavy on disk space framework for Hurl Settings Application. Not sure if it's a good idea or possible to do that easily yet

---

Not sure how u could help with this cause I am not good enough to do divide and conquer. For now, Feedback on the progress would be helpful as useful as other stuff to make better judgements since I'm mostly in dark with my decisions. 

A dynamic rendered inputs would be a good idea for the initial, but not sure how I could do that. anything more on that ??

**Comment by U-C-S** [comment]
KNOWN Issues: (will be fixed in stable version)
- (Fixed) URL edit window makes the hurl minimize the entire window into tray, making it impossible to actually edit URL
  - if it happens, just reload or reopen the application through Taskbar's Tray
- `AutoRoutingRules` property when is null or doesn't exist in UserSettings.json, causes Hurl to crash or not open at all. Case when replacing the v0.8-alpha's exe over previous versions
  - temp workaround: https://github.com/U-C-S/Hurl/discussions/69#discussioncomment-5611897

**Comment by riotrah** [comment]
I can move this to issues if needed, but replacing my exe breaks hurl. 

It simply doesn't appear (including no sys tray presence), and while I do sometimes see it initially in task manager etc, I notice that after some time (haven't tracked), I no longer see any processes running there either.

**Reply by U-C-S** [reply]
😬 oopsie....  As a workaround. add the following property to the UserSettings.json.

```json
  "AutoRoutingRules": []
```

Seems I forgot to handle `null` value case for this property

**Reply by riotrah** [reply]
Oh wow that worked! Thanks for addressing so quickly

**Comment by HankLloydRight** [comment]
I'm just trying to get a simple rule working. 

I just want all URLs with the format "*domain.com*" to go to Vivaldi.

I did have to manually add Vivadli to the UserSettings.json file, but I can't get a simple rule to work. Here's what I have:

```
  "AutoRoutingRules": [
    {
      "Rules": [
        "r$.domain\\.com."
      ],
      "BrowserName": "Vivaldi"
    } 
  ]
```

But it only opens in the default browser and not Vivaldi.  And  I have restarted and reloaded Hurl.
thanks.

**Reply by U-C-S** [reply]
You might have change the regex. Test your regex here - https://regex101.com/

Glob might be more suitable for this - something like `g$*.domain.com/*` for all URLs like `https://meow.domain.com/cat-pics`..

**Comment by HankLloydRight** [comment]
I ended up just asking ChatGPT for the right regex, and it's now working.
I want all links to domain.com to go to  Vivaldi, and all other links to go to Chrome.

```
  "AutoRoutingRules": [
    {
      "Rules": [
        "r$^(?!.*\\bdomain\\.com\\b).*"
      ],
      "BrowserName": "Google Chrome"
    },
    {
      "Rules": [
        "r$\\bdomain\\.com\\b"
      ],
      "BrowserName": "Vivaldi"
    }
  ]
```
